### PR TITLE
Refactor AzureSignalRClient

### PR DESCRIPTION
**Pull request by Y-Sindo**
So that we can reuse `AzureSignalRClient` when we want to negotiation on a specified endpoint.
```cs
var azureSignalRClient = new AzureSignalRClient(serviceHubContext.WithEndpoint(...));
```

### CODE REVIEW

File: src/SignalRServiceExtension/Utils.cs [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
 
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Management;
+
 namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
 {
     internal class Utils
     {
-        public static AzureSignalRClient GetAzureSignalRClient(string connectionStringKey, string attributeHubName)
+        public async static Task<AzureSignalRClient> GetAzureSignalRClient(string connectionStringKey, string attributeHubName)
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/Utils.cs:11]: >GetAzureSignalRClient [](start = 53, length = 21)

GetAzureSignalRClientAsync

File: test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs [hunk 2]
```diff
@@ -38,7 +38,8 @@ public async Task GetClientConnectionInfo()
             var configDict = new Dictionary<string, string>() { { Constants.ServiceTransportTypeName, "Transient" }, { connectionStringKey, connectionString } };
             var configuration = new ConfigurationBuilder().AddInMemoryCollection(configDict).Build();
             var serviceManagerStore = new ServiceManagerStore(configuration, NullLoggerFactory.Instance, new TestRouter());
-            var azureSignalRClient = new AzureSignalRClient(serviceManagerStore, connectionStringKey, hubName);
+            StaticServiceHubContextStore.ServiceManagerStore = serviceManagerStore;
+            var azureSignalRClient = await Microsoft.Azure.WebJobs.Extensions.SignalRService.Utils.GetAzureSignalRClient(connectionStringKey, hubName);
```
Comment (wanlwanl) [inline test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs:42]: >Microsoft.Azure.WebJobs.Extensions.SignalRService [](start = 43, length = 49)

too long, put it as `using ...`

File: test/SignalRServiceExtension.Tests/JobhostEndToEnd.cs [hunk 3]
```diff
@@ -120,15 +120,15 @@ public async Task SignalRAttribute_MissingConnectionStringSettingFacts()
         {
             var task = CreateTestTask(typeof(SignalRFunctions), null);
             var exception = await Assert.ThrowsAsync<FunctionInvocationException>(() => task);
-            Assert.Equal(ErrorMessages.EmptyConnectionStringErrorMessageFormat, exception.InnerException.Message);
```
Comment (wanlwanl) [inline test/SignalRServiceExtension.Tests/JobhostEndToEnd.cs:123]: why change it?

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollectorBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
+{
+    internal class SignalRAsyncCollectorBuilder<T> : IAsyncConverter<SignalRAttribute, IAsyncCollector<T>>
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/Bindings/SignalRAsyncCollectorBuilder.cs:9]: >SignalRAsyncCollectorBuilder [](start = 19, length = 28)

revert?

File: test/SignalRServiceExtension.Tests/JobhostEndToEnd.cs (hunk 3 above)
Comment (Y-Sindo) [reply test/SignalRServiceExtension.Tests/JobhostEndToEnd.cs:123]: The error hierarchy has changed. Hard coding the exception hierarchy is not resilient to changes.

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollectorBuilder.cs (hunk 4 above)
Comment (Y-Sindo) [reply src/SignalRServiceExtension/Bindings/SignalRAsyncCollectorBuilder.cs:9]: Why?

Comment (Y-Sindo) [reply src/SignalRServiceExtension/Bindings/SignalRAsyncCollectorBuilder.cs:9]: This is a new class. The original class is without `Async`.

File: test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs (hunk 2 above)
Comment (Y-Sindo) [reply test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs]: There is another  `Utils`, so we can not use `using ...Utils.GetAzureSignalRClient(...)` https://github.com/Azure/azure-functions-signalrservice-extension/blob/48f279e29f1d9234b95e59ea019e4519af95775f/test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs#L17

Comment (wanlwanl) [reply test/SignalRServiceExtension.Tests/AzureSignalRClientTests.cs]: using SignalRUtils = ....

---
In reply to: [603017311](https://github.com/Azure/azure-functions-signalrservice-extension/pull/193#discussion_r603017311) [](ancestors = 603017311)

**Review by wanlwanl — APPROVED** [review]

**Pull request merged**